### PR TITLE
V2: fix cross validation in `data_split`

### PR DESCRIPTION
**Pull request by hwpang**
## Description
From discussion in https://github.com/chemprop/chemprop/issues/581#issuecomment-1911008302, I realize that the implementation of cross validation in `data_split` is not one usually expects. I make this PR to start a conversation on how we should implement it. @JacksonBurns please let me know if you think this makes sense. 

## Relevant issues
This should close https://github.com/chemprop/chemprop/issues/609.

### CODE REVIEW

**Comment by JacksonBurns** [comment]
I actually think this should go in as-is.

As was alluded to in #524 I had (at one point) greater ambitions for cross-validation - the 'best' design would be something that just calls one of the random sampler multiple times, rather than 'cross validation' being its own kind of sampler. But since `astartes` doesn't implement CV I think this is the most reasonable way forward.

There's a separate discussion to be had (not on the Chemprop repo) about repetitions vs cross validation and _why_ `astartes` doesn't support CV, but yeah, another time and another place.

File: chemprop/data/splitting.py [hunk 1]
```diff
@@ -94,9 +94,24 @@ def split_data(
 
             # returns nested lists of indices
             train, val, test = [], [], []
-            for _ in range(num_folds):
-                result = split_fun(np.arange(len(datapoints)), sampler="random", **astartes_kwargs)
-                i_train, i_val, i_test = _unpack_astartes_result(result, include_val)
+            random = Random(seed)
```
Comment (davidegraff) [inline chemprop/data/splitting.py:97]:
```suggestion
            random = np.random.default_rng(seed)
```

File: chemprop/data/splitting.py [hunk 2]
```diff
@@ -94,9 +94,24 @@ def split_data(
...
+
+            indices = np.tile(np.arange(num_folds), 1 + len(datapoints) // num_folds)[:len(datapoints)]
+            random.shuffle(indices)
+
+            for fold_idx in range(num_folds):
+                test_index = fold_idx
+                val_index = fold_idx + 1 if fold_idx + 1 < num_folds else 0
```
Comment (davidegraff) [inline chemprop/data/splitting.py:104]:
```suggestion
                val_index = (fold_idx + 1) % num_folds
```

**Review by davidegraff — APPROVED** [review]
deferring to @JacksonBurns 's opinion here. Just a minor stylistic suggestions, but you can take or leave. One Q for @JacksonBurns : does astartes use python or numpy for randomness on the backend? We probably want to be consistent in our source here.

**Comment by JacksonBurns** [comment]
> does astartes use python or numpy for randomness on the backend? We probably want to be consistent in our source here.

numpy!

**Pull request merged**